File: examples/python/ping_pong.py
#!/usr/bin/env python3
"""
Ping-Pong Example - Simplest Actor Communication

Usage: python examples/python/ping_pong.py
"""

import asyncio
import pulsing as pul


@pul.remote
class PingPong:
    def ping(self) -> str:
        return "pong"

    def echo(self, msg: str) -> str:
        return f"echo: {msg}"


async def main():
    await pul.init()
    try:
        actor = await PingPong.spawn()

        print(await actor.ping())  # -> pong
        print(await actor.echo("hello"))  # -> echo: hello

        await asyncio.sleep(1)  # Allow background tasks to complete
    finally:
        await pul.shutdown()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/rust/ping_pong.rs
//! Ping-Pong Example - Simplest Actor Communication
//!
//! Run: cargo run --example ping_pong -p pulsing-actor

use pulsing_actor::prelude::*;

struct Echo;

#[async_trait]
impl Actor for Echo {
    async fn receive(&mut self, msg: Message, _ctx: &mut ActorContext) -> anyhow::Result<Message> {
        let s: String = msg.unpack()?;
        Message::pack(&format!("echo: {}", s))
    }
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let system = ActorSystem::builder().build().await?;
    let echo = system.spawn("echo", Echo).await?;

    // ask: send and wait for response
    let resp: String = echo.ask("hello".to_string()).await?;
    println!("{}", resp);

    // tell: fire and forget
    echo.tell("world".to_string()).await?;

    system.shutdown().await
}
